nm: Support backend specific configuration

cathay4t opened the pull request:
Introducing backend specific configuration allowing user to define
specific settings in nmstate schema.

The `get_json_value_difference()` has been modified to accept
stringlized boolean or integer.

Add support of these NM configures:
    * connection.id
    * connection.uuid
    * ipv4.may-fail
    * ipv6.may-fail

Example yaml:

```yaml
---
interfaces:
  - name: eth1
    type: ethernet
    backend-specific:
      networkmanager:
        connection.id: WAN
        ipv4.may-fail: false
    ipv4:
      enabled: true
      dhcp: true
```

Unit test case included to check `get_json_value_difference()`.
Integration test cased included to test NM specific settings, one of
test case is marked as tier 1 due as it is required by openshift
machine-config-operator[1].

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2082042

qinqon commented:
Maybe we can simplify the syntax by just passing directly the attributes that are unknown so we have 

```yaml
---
interfaces:
  - name: eth1
    type: ethernet
    ipv4:
      may-fail: no
      enabled: true
      dhcp: true
    connection:
       id: WAN
```

If they are not part of the schema they are passed to the backend directly, so there is no need for explicit backend stuff at the interface.

Or if we want to be explicit falt it out

```yaml
---
interfaces:
  - name: eth1
    type: ethernet
    ipv4:
      enabled: true
      dhcp: true
    nm.ipv4.may-fail: no
    nm.connection.id: WAN
```

qinqon commented:
> This config will not be serialized and hence not been verified, we are
providing best effort for them.

Why it cannot be serialized and checked ?

cathay4t commented:
> > This config will not be serialized and hence not been verified, we are
> > providing best effort for them.
> 
> Why it cannot be serialized and checked ?

We can. Let me think about it.

cathay4t commented:
New patch uploaded. The backend specific setting will be verified also now.

qinqon commented:
@cathay4t it allow to pass other flags that are not the ones on the PR description ? the idea is to not limit it.

cathay4t commented:
@qinqon The dbus interface of NetworkManger require property type(array, u32, i32, i64,u64, bool), so we cannot allow arbitrary properties. I can expand more if you need anything else.

cathay4t committed:
nm: Support backend specific configuration

Introducing backend specific configuration allowing user to define
specific settings in nmstate schema.

The `get_json_value_difference()` has been modified to accept
stringlized boolean or integer.

Add support of these NM configures:
    * connection.id
    * connection.uuid
    * ipv4.may-fail
    * ipv6.may-fail

Example yaml:

```yaml
---
interfaces:
  - name: eth1
    type: ethernet
    backend-specific:
      networkmanager:
        connection.id: WAN
        ipv4.may-fail: false
    ipv4:
      enabled: true
      dhcp: true
```

Unit test case included to check `get_json_value_difference()`.
Integration test cased included to test NM specific settings, one of
test case is marked as tier 1 due as it is required by openshift
machine-config-operator[1].

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2082042

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
We decided to take another approach to solve this `ipv4.may-fail` opton of NM.
